File: apps/server/src/routes/agents.ts
import type { FastifyInstance } from "fastify";
import { db } from "../db/index.js";
import { agents, conversations, conversationMembers, messages, channels } from "../db/schema.js";
import { eq, and, sql, desc, inArray } from "drizzle-orm";
import { requireAuth } from "../middleware/auth.js";
import {
  createAgentSchema,
  updateAgentSchema,
  pairingExchangeSchema,
} from "@arinova/shared/schemas";
import { generateSecretToken } from "../utils/pairing-code.js";
import { mkdir, writeFile } from "fs/promises";
import path from "path";
import { env } from "../env.js";
import { uploadToR2 } from "../lib/r2.js";

export async function agentRoutes(app: FastifyInstance) {
  // Exchange bot token for agent connection (public — no auth required)
  // MUST be registered before /api/agents/:id to avoid route conflicts
  app.post("/api/agents/pair", async (request, reply) => {
    const body = pairingExchangeSchema.parse(request.body);

    const [agent] = await db
      .select({ id: agents.id, name: agents.name })
      .from(agents)
      .where(eq(agents.secretToken, body.botToken));

    if (!agent) {
      return reply.status(404).send({ error: "Invalid bot token" });
    }

    // Update a2aEndpoint if provided
    if (body.a2aEndpoint) {
      await db
        .update(agents)
        .set({ a2aEndpoint: body.a2aEndpoint, updatedAt: new Date() })
        .where(eq(agents.id, agent.id));
    }

    // Build WS URL for the agent to connect back
    const host = request.headers.host ?? "localhost:3501";
    const protocol = request.headers["x-forwarded-proto"] === "https" ? "wss" : "ws";
    const wsUrl = `${protocol}://${host}/ws/agent`;

    return reply.send({ agentId: agent.id, name: agent.name, wsUrl });
  });

  // Create agent
  app.post("/api/agents", async (request, reply) => {
    const user = await requireAuth(request, reply);
    const body = createAgentSchema.parse(request.body);

    const secretToken = generateSecretToken();

    const [agent] = await db
      .insert(agents)
      .values({
        name: body.name,
        description: body.description ?? null,
        a2aEndpoint: body.a2aEndpoint ?? null,
        secretToken,
        ownerId: user.id,
      })
      .returning();

    return reply.status(201).send(agent);
  });

  // List user's agents
  app.get("/api/agents", async (request, reply) => {
    const user = await requireAuth(request, reply);

    const result = await db
      .select()
      .from(agents)
      .where(eq(agents.ownerId, user.id))
      .orderBy(agents.createdAt);

    return reply.send(result);
  });

  // Get single agent
  app.get<{ Params: { id: string } }>(
    "/api/agents/:id",
    async (request, reply) => {
      const user = await requireAuth(request, reply);

      const [agent] = await db
        .select()
        .from(agents)
        .where(and(eq(agents.id, request.params.id), eq(agents.ownerId, user.id)));

      if (!agent) {
        return reply.status(404).send({ error: "Agent not found" });
      }

      return reply.send(agent);
    }
  );

  // Update agent
  app.put<{ Params: { id: string } }>(
    "/api/agents/:id",
    async (request, reply) => {
      const user = await requireAuth(request, reply);
      const body = updateAgentSchema.parse(request.body);

      const [agent] = await db
        .update(agents)
        .set({ ...body, updatedAt: new Date() })
        .where(and(eq(agents.id, request.params.id), eq(agents.ownerId, user.id)))
        .returning();

      if (!agent) {
        return reply.status(404).send({ error: "Agent not found" });
      }

      return reply.send(agent);
    }
  );

  // Get agent skills from A2A card
  app.get<{ Params: { id: string } }>(
    "/api/agents/:id/skills",
    async (request, reply) => {
      const user = await requireAuth(request, reply);

      const [agent] = await db
        .select({ a2aEndpoint: agents.a2aEndpoint })
        .from(agents)
        .where(and(eq(agents.id, request.params.id), eq(agents.ownerId, user.id)));

      if (!agent) {
        return reply.status(404).send({ error: "Agent not found" });
      }

      if (!agent.a2aEndpoint) {
        return reply.send({ skills: [] });
      }

      try {
        const cardUrl = agent.a2aEndpoint;
        const res = await fetch(cardUrl, {
          headers: { Accept: "application/json" },
          signal: AbortSignal.timeout(5000),
        });

        if (!res.ok) {
          return reply.send({ skills: [] });
        }

        const card = (await res.json()) as {
          skills?: { id: string; name: string; description?: string }[];
        };

        const skills = (card.skills ?? []).map((s) => ({
          id: s.id,
          name: s.name,
          description: s.description ?? "",
        }));

        return reply.send({ skills });
      } catch {
        return reply.send({ skills: [] });
      }
    }
  );

  // Upload agent avatar
  app.post<{ Params: { id: string } }>(
    "/api/agents/:id/avatar",
    async (request, reply) => {
      const user = await requireAuth(request, reply);

      // Verify ownership
      const [agent] = await db
        .select({ id: agents.id })
        .from(agents)
        .where(and(eq(agents.id, request.params.id), eq(agents.ownerId, user.id)));

      if (!agent) {
        return reply.status(404).send({ error: "Agent not found" });
      }

      const data = await request.file();
      if (!data) {
        return reply.status(400).send({ error: "No file uploaded" });
      }

      // Validate image type
      if (!data.mimetype.startsWith("image/")) {
        return reply.status(400).send({ error: "Only image files are allowed" });
      }

      const buffer = await data.toBuffer();

      // Max 2MB for avatars
      if (buffer.length > 2 * 1024 * 1024) {
        return reply.status(400).send({ error: "Avatar must be under 2MB" });
      }

      // Save file
      const ext = data.filename.split(".").pop() ?? "jpg";
      const filename = `avatar_${agent.id}_${Date.now()}.${ext}`;
      const r2Key = `avatars/${filename}`;

      // Upload to R2 if configured, otherwise local disk
      const r2Url = await uploadToR2(r2Key, buffer, data.mimetype);
      let avatarUrl: string;

      if (r2Url) {
        avatarUrl = r2Url;
      } else {
        const avatarDir = path.resolve(env.UPLOAD_DIR, "avatars");
        await mkdir(avatarDir, { recursive: true });
        await writeFile(path.join(avatarDir, filename), buffer);
        avatarUrl = `/uploads/avatars/${filename}`;
      }

      // Update agent
      const [updated] = await db
        .update(agents)
        .set({ avatarUrl, updatedAt: new Date() })
        .where(eq(agents.id, agent.id))
        .returning();

      return reply.send({ avatarUrl: updated.avatarUrl });
    }
  );

  // Regenerate bot secret token (old token is immediately invalidated)
  app.post<{ Params: { id: string } }>(
    "/api/agents/:id/regenerate-token",
    async (request, reply) => {
      const user = await requireAuth(request, reply);

      const [agent] = await db
        .select({ id: agents.id })
        .from(agents)
        .where(and(eq(agents.id, request.params.id), eq(agents.ownerId, user.id)));

      if (!agent) {
        return reply.status(404).send({ error: "Agent not found" });
      }

      const secretToken = generateSecretToken();

      await db
        .update(agents)
        .set({ secretToken, updatedAt: new Date() })
        .where(eq(agents.id, agent.id));

      return reply.send({ secretToken });
    }
  );

  // Get agent usage stats
  app.get<{ Params: { id: string } }>(
    "/api/agents/:id/stats",
    async (request, reply) => {
      const user = await requireAuth(request, reply);

      const [agent] = await db
        .select({ id: agents.id })
        .from(agents)
        .where(and(eq(agents.id, request.params.id), eq(agents.ownerId, user.id)));

      if (!agent) {
        return reply.status(404).send({ error: "Agent not found" });
      }

      // Get conversation IDs for this agent
      const agentConvos = await db
        .select({ id: conversations.id })
        .from(conversations)
        .where(and(eq(conversations.agentId, agent.id), eq(conversations.userId, user.id)));

      const convoIds = agentConvos.map((c) => c.id);

      let totalMessages = 0;
      let lastActive: Date | null = null;

      if (convoIds.length > 0) {
        const [msgStats] = await db
          .select({
            count: sql<number>`count(*)::int`,
            lastMessage: sql<Date>`max(${messages.createdAt})`,
          })
          .from(messages)
          .where(sql`${messages.conversationId} = ANY(${convoIds})`);

        totalMessages = msgStats?.count ?? 0;
        lastActive = msgStats?.lastMessage ?? null;
      }

      return reply.send({
        totalMessages,
        totalConversations: convoIds.length,
        lastActive,
      });
    }
  );

  // Clear all conversation history for an agent
  app.delete<{ Params: { id: string } }>(
    "/api/agents/:id/history",
    async (request, reply) => {
      const user = await requireAuth(request, reply);

      const [agent] = await db
        .select({ id: agents.id })
        .from(agents)
        .where(and(eq(agents.id, request.params.id), eq(agents.ownerId, user.id)));

      if (!agent) {
        return reply.status(404).send({ error: "Agent not found" });
      }

      // Delete all conversations with this agent (messages cascade)
      await db
        .delete(conversations)
        .where(and(eq(conversations.agentId, agent.id), eq(conversations.userId, user.id)));

      return reply.status(204).send();
    }
  );

  // Export chat history for an agent
  app.get<{ Params: { id: string }; Querystring: { format?: string } }>(
    "/api/agents/:id/export",
    async (request, reply) => {
      const user = await requireAuth(request, reply);
      const format = (request.query as { format?: string }).format ?? "json";

      const [agent] = await db
        .select({ id: agents.id, name: agents.name })
        .from(agents)
        .where(and(eq(agents.id, request.params.id), eq(agents.ownerId, user.id)));

      if (!agent) {
        return reply.status(404).send({ error: "Agent not found" });
      }

      const convos = await db
        .select()
        .from(conversations)
        .where(and(eq(conversations.agentId, agent.id), eq(conversations.userId, user.id)))
        .orderBy(conversations.createdAt);

      const result = [];
      for (const convo of convos) {
        const msgs = await db
          .select()
          .from(messages)
          .where(eq(messages.conversationId, convo.id))
          .orderBy(messages.createdAt);

        result.push({
          conversationId: convo.id,
          title: convo.title,
          createdAt: convo.createdAt,
          messages: msgs.map((m) => ({
            role: m.role,
            content: m.content,
            status: m.status,
            createdAt: m.createdAt,
          })),
        });
      }

      if (format === "markdown") {
        let md = `# Chat Export: ${agent.name}\n\n`;
        for (const convo of result) {
          md += `## ${convo.title ?? "Untitled"} (${new Date(convo.createdAt).toLocaleDateString()})\n\n`;
          for (const msg of convo.messages) {
            const role = msg.role === "user" ? "You" : agent.name;
            const time = new Date(msg.createdAt).toLocaleTimeString();
            md += `**${role}** [${time}]\n${msg.content}\n\n`;
          }
          md += "---\n\n";
        }
        reply.header("Content-Type", "text/markdown");
        reply.header("Content-Disposition", `attachment; filename="${agent.name}-export.md"`);
        return reply.send(md);
      }

      reply.header("Content-Type", "application/json");
      reply.header("Content-Disposition", `attachment; filename="${agent.name}-export.json"`);
      return reply.send({ agent: agent.name, exportedAt: new Date(), conversations: result });
    }
  );

  // Delete agent
  app.delete<{ Params: { id: string } }>(
    "/api/agents/:id",
    async (request, reply) => {
      const user = await requireAuth(request, reply);

      // Verify ownership first
      const [agent] = await db
        .select({ id: agents.id })
        .from(agents)
        .where(and(eq(agents.id, request.params.id), eq(agents.ownerId, user.id)));

      if (!agent) {
        return reply.status(404).send({ error: "Agent not found" });
      }

      const agentId = agent.id;

      // Clean up all references before deleting the agent:
      // 1. Remove from group conversation members
      await db.delete(conversationMembers).where(eq(conversationMembers.agentId, agentId));

      // 2. Delete messages in direct conversations with this agent, then the conversations
      const directConvos = await db
        .select({ id: conversations.id })
        .from(conversations)
        .where(eq(conversations.agentId, agentId));

      if (directConvos.length > 0) {
        const convIds = directConvos.map((c) => c.id);
        await db.delete(messages).where(inArray(messages.conversationId, convIds));
        await db.delete(conversations).where(inArray(conversations.id, convIds));
      }

      // 3. Unlink from community channels (nullable FK, set to null)
      await db.update(channels).set({ agentId: null }).where(eq(channels.agentId, agentId));

      // 4. Delete the agent
      await db.delete(agents).where(eq(agents.id, agentId));

      return reply.status(204).send();
    }
  );
}

File: apps/rust-server/src/routes/agents.rs
use axum::{
    extract::{Multipart, Path, Query, State},
    http::StatusCode,
    response::{IntoResponse, Json, Response},
    routing::{delete, get, post, put},
    Router,
};
use serde::Deserialize;
use serde_json::{json, Value};
use uuid::Uuid;

use crate::auth::middleware::AuthUser;
use crate::db::models::Agent;
use crate::utils::pairing_code::generate_secret_token;
use crate::ws::state::WsState;
use crate::AppState;

pub fn router() -> Router<AppState> {
    Router::new()
        .route("/api/agents", post(create_agent).get(list_agents))
        .route(
            "/api/agents/{id}",
            get(get_agent).put(update_agent).delete(delete_agent),
        )
        .route("/api/agents/{id}/skills", get(get_skills))
        .route("/api/agents/{id}/avatar", post(upload_avatar))
        .route("/api/agents/{id}/regenerate-token", post(regenerate_token))
        .route("/api/agents/{id}/stats", get(get_stats))
        .route("/api/agents/{id}/history", delete(clear_history))
        .route("/api/agents/{id}/export", get(export_history))
}

#[derive(Deserialize)]
struct CreateAgentBody {
    name: String,
    description: Option<String>,
    #[serde(rename = "a2aEndpoint")]
    a2a_endpoint: Option<String>,
}

#[derive(Deserialize)]
struct UpdateAgentBody {
    name: Option<String>,
    description: Option<String>,
    #[serde(rename = "a2aEndpoint")]
    a2a_endpoint: Option<String>,
    #[serde(rename = "systemPrompt")]
    system_prompt: Option<String>,
    #[serde(rename = "welcomeMessage")]
    welcome_message: Option<String>,
    #[serde(rename = "quickReplies")]
    quick_replies: Option<serde_json::Value>,
    #[serde(rename = "notificationsEnabled")]
    notifications_enabled: Option<bool>,
    #[serde(rename = "isPublic")]
    is_public: Option<bool>,
    category: Option<String>,
}

#[derive(Deserialize)]
struct ExportQuery {
    format: Option<String>,
}

async fn create_agent(
    State(state): State<AppState>,
    user: AuthUser,
    Json(body): Json<CreateAgentBody>,
) -> Response {
    let token = generate_secret_token();

    let result = sqlx::query_as::<_, Agent>(
        r#"INSERT INTO agents (name, description, a2a_endpoint, secret_token, owner_id)
           VALUES ($1, $2, $3, $4, $5)
           RETURNING *"#,
    )
    .bind(&body.name)
    .bind(&body.description)
    .bind(&body.a2a_endpoint)
    .bind(&token)
    .bind(&user.id)
    .fetch_one(&state.db)
    .await;

    match result {
        Ok(agent) => (StatusCode::CREATED, Json(json!(agent))).into_response(),
        Err(e) => (
            StatusCode::INTERNAL_SERVER_ERROR,
            Json(json!({"error": e.to_string()})),
        )
            .into_response(),
    }
}

async fn list_agents(State(state): State<AppState>, user: AuthUser) -> Response {
    let agents = sqlx::query_as::<_, Agent>(
        "SELECT * FROM agents WHERE owner_id = $1 ORDER BY created_at",
    )
    .bind(&user.id)
    .fetch_all(&state.db)
    .await;

    match agents {
        Ok(agents) => Json(json!(agents)).into_response(),
        Err(e) => (
            StatusCode::INTERNAL_SERVER_ERROR,
            Json(json!({"error": e.to_string()})),
        )
            .into_response(),
    }
}

async fn get_agent(
    State(state): State<AppState>,
    user: AuthUser,
    Path(id): Path<Uuid>,
) -> Response {
    let agent = sqlx::query_as::<_, Agent>(
        "SELECT * FROM agents WHERE id = $1 AND owner_id = $2",
    )
    .bind(id)
    .bind(&user.id)
    .fetch_optional(&state.db)
    .await;

    match agent {
        Ok(Some(agent)) => Json(json!(agent)).into_response(),
        Ok(None) => (
            StatusCode::NOT_FOUND,
            Json(json!({"error": "Agent not found"})),
        )
            .into_response(),
        Err(e) => (
            StatusCode::INTERNAL_SERVER_ERROR,
            Json(json!({"error": e.to_string()})),
        )
            .into_response(),
    }
}

async fn update_agent(
    State(state): State<AppState>,
    user: AuthUser,
    Path(id): Path<Uuid>,
    Json(body): Json<UpdateAgentBody>,
) -> Response {
    // Build dynamic update query
    let mut sets = vec!["updated_at = NOW()".to_string()];
    let mut param_idx = 3u32; // $1 = id, $2 = owner_id

    // We'll use a simpler approach - update all provided fields
    let result = sqlx::query_as::<_, Agent>(
        r#"UPDATE agents SET
           name = COALESCE($3, name),
           description = COALESCE($4, description),
           a2a_endpoint = COALESCE($5, a2a_endpoint),
           system_prompt = COALESCE($6, system_prompt),
           welcome_message = COALESCE($7, welcome_message),
           quick_replies = COALESCE($8, quick_replies),
           notifications_enabled = COALESCE($9, notifications_enabled),
           is_public = COALESCE($10, is_public),
           category = COALESCE($11, category),
           updated_at = NOW()
           WHERE id = $1 AND owner_id = $2
           RETURNING *"#,
    )
    .bind(id)
    .bind(&user.id)
    .bind(&body.name)
    .bind(&body.description)
    .bind(&body.a2a_endpoint)
    .bind(&body.system_prompt)
    .bind(&body.welcome_message)
    .bind(&body.quick_replies)
    .bind(body.notifications_enabled)
    .bind(body.is_public)
    .bind(&body.category)
    .fetch_optional(&state.db)
    .await;

    match result {
        Ok(Some(agent)) => Json(json!(agent)).into_response(),
        Ok(None) => (
            StatusCode::NOT_FOUND,
            Json(json!({"error": "Agent not found"})),
        )
            .into_response(),
        Err(e) => (
            StatusCode::INTERNAL_SERVER_ERROR,
            Json(json!({"error": e.to_string()})),
        )
            .into_response(),
    }
}

async fn get_skills(
    State(state): State<AppState>,
    user: AuthUser,
    Path(id): Path<Uuid>,
) -> Response {
    let agent = sqlx::query_as::<_, (Uuid, Option<String>)>(
        "SELECT id, a2a_endpoint FROM agents WHERE id = $1 AND owner_id = $2",
    )
    .bind(id)
    .bind(&user.id)
    .fetch_optional(&state.db)
    .await;

    let (agent_id, a2a_endpoint) = match agent {
        Ok(Some(a)) => a,
        Ok(None) => {
            return (
                StatusCode::NOT_FOUND,
                Json(json!({"error": "Agent not found"})),
            )
                .into_response();
        }
        Err(e) => {
            return (
                StatusCode::INTERNAL_SERVER_ERROR,
                Json(json!({"error": e.to_string()})),
            )
                .into_response();
        }
    };

    // Check WS-declared skills
    let ws_skills = state.ws.get_agent_skills(&agent_id.to_string());
    if !ws_skills.is_empty() {
        return Json(json!({"skills": ws_skills})).into_response();
    }

    // Fallback: fetch from A2A card
    if let Some(endpoint) = a2a_endpoint {
        if let Ok(resp) = reqwest::Client::new()
            .get(&endpoint)
            .header("Accept", "application/json")
            .timeout(std::time::Duration::from_secs(5))
            .send()
            .await
        {
            if resp.status().is_success() {
                if let Ok(card) = resp.json::<Value>().await {
                    if let Some(skills) = card.get("skills").and_then(|s| s.as_array()) {
                        let mapped: Vec<Value> = skills
                            .iter()
                            .map(|s| {
                                json!({
                                    "id": s.get("id").and_then(|v| v.as_str()).unwrap_or(""),
                                    "name": s.get("name").and_then(|v| v.as_str()).unwrap_or(""),
                                    "description": s.get("description").and_then(|v| v.as_str()).unwrap_or(""),
                                })
                            })
                            .collect();
                        return Json(json!({"skills": mapped})).into_response();
                    }
                }
            }
        }
    }

    Json(json!({"skills": []})).into_response()
}

async fn upload_avatar(
    State(state): State<AppState>,
    user: AuthUser,
    Path(id): Path<Uuid>,
    mut multipart: Multipart,
) -> Response {
    // Verify ownership
    let agent = sqlx::query_as::<_, (Uuid,)>(
        "SELECT id FROM agents WHERE id = $1 AND owner_id = $2",
    )
    .bind(id)
    .bind(&user.id)
    .fetch_optional(&state.db)
    .await;

    if matches!(agent, Ok(None) | Err(_)) {
        return (
            StatusCode::NOT_FOUND,
            Json(json!({"error": "Agent not found"})),
        )
            .into_response();
    }

    while let Ok(Some(field)) = multipart.next_field().await {
        let content_type = field.content_type().unwrap_or("").to_string();
        if !content_type.starts_with("image/") {
            return (
                StatusCode::BAD_REQUEST,
                Json(json!({"error": "Only image files are allowed"})),
            )
                .into_response();
        }

        let filename = field
            .file_name()
            .unwrap_or("avatar.jpg")
            .to_string();
        let data = match field.bytes().await {
            Ok(d) => d,
            Err(_) => {
                return (
                    StatusCode::BAD_REQUEST,
                    Json(json!({"error": "Failed to read file"})),
                )
                    .into_response();
            }
        };

        if data.len() > 2 * 1024 * 1024 {
            return (
                StatusCode::BAD_REQUEST,
                Json(json!({"error": "Avatar must be under 2MB"})),
            )
                .into_response();
        }

        let ext = filename.rsplit('.').next().unwrap_or("jpg");
        let stored = format!("avatar_{}_{}.{}", id, chrono::Utc::now().timestamp(), ext);
        let r2_key = format!("avatars/{}", stored);

        let avatar_url = if let Some(s3) = &state.s3 {
            match crate::services::r2::upload_to_r2(s3, &state.config.r2_bucket, &r2_key, data.to_vec(), &content_type, &state.config.r2_public_url).await {
                Ok(url) => url,
                Err(_) => format!("/uploads/avatars/{}", stored),
            }
        } else {
            let dir = std::path::Path::new(&state.config.upload_dir).join("avatars");
            let _ = tokio::fs::create_dir_all(&dir).await;
            let _ = tokio::fs::write(dir.join(&stored), &data).await;
            format!("/uploads/avatars/{}", stored)
        };

        let _ = sqlx::query(
            "UPDATE agents SET avatar_url = $1, updated_at = NOW() WHERE id = $2",
        )
        .bind(&avatar_url)
        .bind(id)
        .execute(&state.db)
        .await;

        return Json(json!({"avatarUrl": avatar_url})).into_response();
    }

    (
        StatusCode::BAD_REQUEST,
        Json(json!({"error": "No file uploaded"})),
    )
        .into_response()
}

async fn regenerate_token(
    State(state): State<AppState>,
    user: AuthUser,
    Path(id): Path<Uuid>,
) -> Response {
    let agent = sqlx::query_as::<_, (Uuid,)>(
        "SELECT id FROM agents WHERE id = $1 AND owner_id = $2",
    )
    .bind(id)
    .bind(&user.id)
    .fetch_optional(&state.db)
    .await;

    if matches!(agent, Ok(None) | Err(_)) {
        return (
            StatusCode::NOT_FOUND,
            Json(json!({"error": "Agent not found"})),
        )
            .into_response();
    }

    let new_token = generate_secret_token();
    let _ = sqlx::query("UPDATE agents SET secret_token = $1, updated_at = NOW() WHERE id = $2")
        .bind(&new_token)
        .bind(id)
        .execute(&state.db)
        .await;

    Json(json!({"secretToken": new_token})).into_response()
}

async fn get_stats(
    State(state): State<AppState>,
    user: AuthUser,
    Path(id): Path<Uuid>,
) -> Response {
    let agent = sqlx::query_as::<_, (Uuid,)>(
        "SELECT id FROM agents WHERE id = $1 AND owner_id = $2",
    )
    .bind(id)
    .bind(&user.id)
    .fetch_optional(&state.db)
    .await;

    if matches!(agent, Ok(None) | Err(_)) {
        return (
            StatusCode::NOT_FOUND,
            Json(json!({"error": "Agent not found"})),
        )
            .into_response();
    }

    let stats = sqlx::query_as::<_, (i64, Option<chrono::NaiveDateTime>)>(
        r#"SELECT COUNT(*)::bigint, MAX(m.created_at)
           FROM messages m
           JOIN conversations c ON m.conversation_id = c.id
           WHERE c.agent_id = $1 AND c.user_id = $2"#,
    )
    .bind(id)
    .bind(&user.id)
    .fetch_one(&state.db)
    .await;

    let conv_count = sqlx::query_as::<_, (i64,)>(
        "SELECT COUNT(*)::bigint FROM conversations WHERE agent_id = $1 AND user_id = $2",
    )
    .bind(id)
    .bind(&user.id)
    .fetch_one(&state.db)
    .await;

    let (total_messages, last_active) = stats.unwrap_or((0, None));
    let total_conversations = conv_count.map(|c| c.0).unwrap_or(0);

    Json(json!({
        "totalMessages": total_messages,
        "totalConversations": total_conversations,
        "lastActive": last_active,
    }))
    .into_response()
}

async fn clear_history(
    State(state): State<AppState>,
    user: AuthUser,
    Path(id): Path<Uuid>,
) -> Response {
    let agent = sqlx::query_as::<_, (Uuid,)>(
        "SELECT id FROM agents WHERE id = $1 AND owner_id = $2",
    )
    .bind(id)
    .bind(&user.id)
    .fetch_optional(&state.db)
    .await;

    if matches!(agent, Ok(None) | Err(_)) {
        return (
            StatusCode::NOT_FOUND,
            Json(json!({"error": "Agent not found"})),
        )
            .into_response();
    }

    let _ = sqlx::query(
        "DELETE FROM conversations WHERE agent_id = $1 AND user_id = $2",
    )
    .bind(id)
    .bind(&user.id)
    .execute(&state.db)
    .await;

    StatusCode::NO_CONTENT.into_response()
}

async fn export_history(
    State(state): State<AppState>,
    user: AuthUser,
    Path(id): Path<Uuid>,
    Query(query): Query<ExportQuery>,
) -> Response {
    let format = query.format.as_deref().unwrap_or("json");

    let agent = sqlx::query_as::<_, (Uuid, String)>(
        "SELECT id, name FROM agents WHERE id = $1 AND owner_id = $2",
    )
    .bind(id)
    .bind(&user.id)
    .fetch_optional(&state.db)
    .await;

    let (_, agent_name) = match agent {
        Ok(Some(a)) => a,
        _ => {
            return (
                StatusCode::NOT_FOUND,
                Json(json!({"error": "Agent not found"})),
            )
                .into_response();
        }
    };

    let convos = sqlx::query_as::<_, crate::db::models::Conversation>(
        "SELECT * FROM conversations WHERE agent_id = $1 AND user_id = $2 ORDER BY created_at",
    )
    .bind(id)
    .bind(&user.id)
    .fetch_all(&state.db)
    .await
    .unwrap_or_default();

    let mut result = Vec::new();
    for convo in &convos {
        let msgs = sqlx::query_as::<_, crate::db::models::Message>(
            "SELECT * FROM messages WHERE conversation_id = $1 ORDER BY created_at",
        )
        .bind(convo.id)
        .fetch_all(&state.db)
        .await
        .unwrap_or_default();

        result.push(json!({
            "conversationId": convo.id,
            "title": convo.title,
            "createdAt": convo.created_at.and_utc().to_rfc3339(),
            "messages": msgs.iter().map(|m| json!({
                "role": m.role,
                "content": m.content,
                "status": m.status,
                "createdAt": m.created_at.and_utc().to_rfc3339(),
            })).collect::<Vec<_>>(),
        }));
    }

    if format == "markdown" {
        let mut md = format!("# Chat Export: {}\n\n", agent_name);
        for convo in &result {
            let title = convo["title"].as_str().unwrap_or("Untitled");
            md.push_str(&format!("## {}\n\n", title));
            if let Some(msgs) = convo["messages"].as_array() {
                for msg in msgs {
                    let role = if msg["role"].as_str() == Some("user") {
                        "You"
                    } else {
                        &agent_name
                    };
                    let content = msg["content"].as_str().unwrap_or("");
                    md.push_str(&format!("**{}**\n{}\n\n", role, content));
                }
            }
            md.push_str("---\n\n");
        }

        let mut resp = md.into_response();
        resp.headers_mut().insert("content-type", "text/markdown".parse().unwrap());
        resp.headers_mut().insert(
            "content-disposition",
            format!("attachment; filename=\"{}-export.md\"", agent_name).parse().unwrap(),
        );
        return resp;
    }

    let mut resp = Json(json!({
        "agent": agent_name,
        "exportedAt": chrono::Utc::now(),
        "conversations": result,
    }))
    .into_response();
    resp.headers_mut().insert("content-type", "application/json".parse().unwrap());
    resp.headers_mut().insert(
        "content-disposition",
        format!("attachment; filename=\"{}-export.json\"", agent_name).parse().unwrap(),
    );
    resp
}

async fn delete_agent(
    State(state): State<AppState>,
    user: AuthUser,
    Path(id): Path<Uuid>,
) -> Response {
    let agent = sqlx::query_as::<_, (Uuid,)>(
        "SELECT id FROM agents WHERE id = $1 AND owner_id = $2",
    )
    .bind(id)
    .bind(&user.id)
    .fetch_optional(&state.db)
    .await;

    if matches!(agent, Ok(None) | Err(_)) {
        return (
            StatusCode::NOT_FOUND,
            Json(json!({"error": "Agent not found"})),
        )
            .into_response();
    }

    // Clean up references
    let _ = sqlx::query("DELETE FROM conversation_members WHERE agent_id = $1")
        .bind(id)
        .execute(&state.db)
        .await;

    let direct_convos = sqlx::query_as::<_, (Uuid,)>(
        "SELECT id FROM conversations WHERE agent_id = $1",
    )
    .bind(id)
    .fetch_all(&state.db)
    .await
    .unwrap_or_default();

    if !direct_convos.is_empty() {
        let conv_ids: Vec<Uuid> = direct_convos.into_iter().map(|c| c.0).collect();
        for conv_id in &conv_ids {
            let _ = sqlx::query("DELETE FROM messages WHERE conversation_id = $1")
                .bind(conv_id)
                .execute(&state.db)
                .await;
            let _ = sqlx::query("DELETE FROM conversations WHERE id = $1")
                .bind(conv_id)
                .execute(&state.db)
                .await;
        }
    }

    let _ = sqlx::query("UPDATE channels SET agent_id = NULL WHERE agent_id = $1")
        .bind(id)
        .execute(&state.db)
        .await;

    let _ = sqlx::query("DELETE FROM agents WHERE id = $1")
        .bind(id)
        .execute(&state.db)
        .await;

    StatusCode::NO_CONTENT.into_response()
}
